src/ovs/aca_vlan_manager.cpp

```diff
@@ -39,60 +41,44 @@ void ACA_Vlan_Manager::clear_all_data()
 {
   ACA_LOG_DEBUG("%s", "ACA_Vlan_Manager::clear_all_data ---> Entering\n");
 
-  auto vpcs_table_mutex_start = chrono::steady_clock::now();
-  // -----critical section starts-----
-  _vpcs_table_mutex.lock();
-  // All the elements in the unordered_map container are dropped:
+  // All the elements in the container are deleted:
   // their destructors are called, and they are removed from the container,
-  // leaving _vpcs_table with a size of 0.
+  // leaving an empty _vpcs_table.
   _vpcs_table.clear();
-  _vpcs_table_mutex.unlock();
-  // -----critical section ends-----
-  auto vpcs_table_mutex_end = chrono::steady_clock::now();
-
-  g_total_vpcs_table_mutex_time +=
-          cast_to_microseconds(vpcs_table_mutex_end - vpcs_table_mutex_start).count();
-
-  // auto vpcs_table_mutex_elapse_time =
-  //         cast_to_microseconds(vpcs_table_mutex_end - vpcs_table_mutex_start).count();
-
-  // g_total_vpcs_table_mutex_time += vpcs_table_mutex_elapse_time;
 
   ACA_LOG_DEBUG("%s", "ACA_Vlan_Manager::clear_all_data <--- Exiting\n");
 }
 
-// unsafe function, needs to be called inside vpc_table_mutex lock
 // this function assumes there is no existing entry for vpc_id
-void ACA_Vlan_Manager::create_entry_unsafe(uint tunnel_id)
+void ACA_Vlan_Manager::create_entry(uint tunnel_id)
 {
-  ACA_LOG_DEBUG("%s", "ACA_Vlan_Manager::create_entry_unsafe ---> Entering\n");
+  ACA_LOG_DEBUG("%s", "ACA_Vlan_Manager::create_entry ---> Entering\n");
 
-  vpc_table_entry new_table_entry;
-  new_table_entry.vlan_id = current_available_vlan_id.load();
-  current_available_vlan_id++;
+  vpc_table_entry *new_vpc_table_entry = new vpc_table_entry;
+  // fetch the value first to used for new_vpc_table_entry->vlan_id
+  // then add 1 after, doing both atomically
+  // std::memory_order_relaxed option won't help much for x86 but other
+  // CPU architecture can take advantage of it
+  new_vpc_table_entry->vlan_id =
+          current_available_vlan_id.fetch_add(1, std::memory_order_relaxed);
 
-  _vpcs_table.emplace(tunnel_id, new_table_entry);
+  _vpcs_table.insert(tunnel_id, new_vpc_table_entry);
```

Comment:
If the key is tunnel_id, do we expect the new hash map help in-VPC performance, e.g., a large vpc with 10,000s ports?

Reply:
I expect the hash function is much easier/faster for hashing integer (tunnel_id) vs string (vpc_id) used previously. With the improvement in hashing function, it will bring improvement to both read and write operations because it all starts with hashing.